From 3c6f6086126ef8901be10db13c60b9c6c03241e0 Mon Sep 17 00:00:00 2001
From: Nikolay Vakhlyarskiy <nvakhlyarskiy@gmail.com>
Date: Tue, 16 Apr 2019 11:50:47 -0700
Subject: [PATCH 0629/1852] v4l: xilinx: dma: add input ioctls

commit 5e5ff34b305c8505802da79c0b46d680559b641d from
https://github.com/Xilinx/linux-xlnx.git

According to :

https://www.linuxtv.org/downloads/v4l-dvb-apis/uapi/v4l/video.html

Drivers must implement all the input ioctls when the device
has one or more inputs, all the output ioctls when the device
has one or more outputs.

As the experience from practice, OpenCV fails to create the
instance of VideoCapture object when there is no implementation
for VIDIOC_ENUMINPUT, VIDIOC_G_INPUT, VIDIOC_S_INPUT IOCTLs
in device driver.

Signed-off-by: Nikolay Vakhlyarskiy <nvakhlyarskiy@gmail.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 44 ++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 35f27a1d5b39..95ef373d9e1e 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -823,6 +823,47 @@ static int xvip_xdma_enum_fmt(struct xvip_dma *dma, struct v4l2_fmtdesc *f,
 	return 0;
 }
 
+static int
+xvip_dma_enum_input(struct file *file, void *priv, struct v4l2_input *i)
+{
+	struct v4l2_fh *vfh = file->private_data;
+	struct xvip_dma *dma = to_xvip_dma(vfh->vdev);
+	struct v4l2_subdev *subdev;
+
+	if (i->index > 0)
+		return -EINVAL;
+
+	subdev = xvip_dma_remote_subdev(&dma->pad, NULL);
+	if (!subdev)
+		return -EPIPE;
+
+	/*
+	 * FIXME: right now only camera input type is handled.
+	 * There should be mechanism to distinguish other types of
+	 * input like V4L2_INPUT_TYPE_TUNER and V4L2_INPUT_TYPE_TOUCH.
+	 */
+	i->type = V4L2_INPUT_TYPE_CAMERA;
+	strlcpy(i->name, subdev->name, sizeof(i->name));
+
+	return 0;
+}
+
+static int
+xvip_dma_get_input(struct file *file, void *fh, unsigned int *i)
+{
+	*i = 0;
+	return 0;
+}
+
+static int
+xvip_dma_set_input(struct file *file, void *fh, unsigned int i)
+{
+	if (i > 0)
+		return -EINVAL;
+
+	return 0;
+}
+
 /* FIXME: without this callback function, some applications are not configured
  * with correct formats, and it results in frames in wrong format. Whether this
  * callback needs to be required is not clearly defined, so it should be
@@ -1120,6 +1161,9 @@ static const struct v4l2_ioctl_ops xvip_dma_ioctl_ops = {
 	.vidioc_expbuf			= vb2_ioctl_expbuf,
 	.vidioc_streamon		= vb2_ioctl_streamon,
 	.vidioc_streamoff		= vb2_ioctl_streamoff,
+	.vidioc_enum_input	= &xvip_dma_enum_input,
+	.vidioc_g_input		= &xvip_dma_get_input,
+	.vidioc_s_input		= &xvip_dma_set_input,
 };
 
 /* -----------------------------------------------------------------------------
-- 
2.31.1

